tools/ExampleTester/FastCsprojCompilationParser.cs

```diff
@@ -0,0 +1,163 @@
+﻿using System.Collections.Frozen;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+
+namespace ExampleTester;
+
+public static class FastCsprojCompilationParser
+{
+    public static CSharpCompilation CreateCompilation(string csprojPath)
+    {
+        var csproj = ParseCsproj(XDocument.Load(csprojPath), csprojPath);
+
+        var syntaxTrees = Directory.GetFiles(Path.GetDirectoryName(csprojPath)!, "*.cs").AsParallel().Select(path =>
+        {
+            using var stream = File.OpenRead(path);
+            return CSharpSyntaxTree.ParseText(SourceText.From(stream), csproj.ParseOptions, path);
+        });
+
+        return CSharpCompilation.Create(
+            csproj.AssemblyName,
+            csproj.GeneratedSources.AddRange(syntaxTrees),
+            SupportedTargetFrameworks[csproj.TargetFramework].References,
+            csproj.CompilationOptions);
+    }
+
+    private static readonly FrozenDictionary<string, OutputKind> SupportedOutputKinds = new KeyValuePair<string, OutputKind>[]
```

Comment:
Any reason for using FrozenDictionary instead of ImmutableDictionary? Mostly out of curiosity, really. (I haven't used the frozen collections, but have used immutable collections a fair amount. From what I've just read, I suspect that most of the time I possibly *should* be using frozen collections...)

Reply:
I don't believe these is a reason not to use them. I still use ImmutableDictionary plenty in places where the extra speed just doesn't matter, just for consistency. It does sound potentially helpful for a web app.
ImmutableList's internal structure doesnt permit random access, yet the class has a indexer which thus performs pretty slowly if you have a loop with repeated indexer usage. Unless there will be frequent updates, I usually stick with ImmutableArray which I've been told is even more favorable than arrays for performance because there's no variance checking needed.